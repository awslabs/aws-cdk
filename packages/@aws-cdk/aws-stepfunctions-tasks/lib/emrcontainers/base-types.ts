/**
 * The information about job driver for Spark submit.
 */
export interface SparkSubmitJobDriver {

  /**
   * The entry point of job application.
   * Length Constraints: Minimum length of 1. Maximum length of 256.
   *
   * @default - No entry point
   */
  readonly entryPoint: string;

  /**
   * The arguments for job application.
   * Length Constraints: Minimum length of 1. Maximum length of 10280.
   *
   * @default - No arguments defined
   */
  readonly entryPointArguments?: string[];

  /**
   * The Spark submit parameters that are used for job runs.
   * Length Constraints: Minimum length of 1. Maximum length of 102400.
   *
   * @default - No spark submit parameters
   */
  readonly sparkSubmitParameters?: string;
}

/**
 * Specify the driver that the job runs on.
 */
export interface JobDriver {

  /**
   * The job driver parameters specified for spark submit.
   *
   * @see https://docs.aws.amazon.com/emr-on-eks/latest/APIReference/API_SparkSubmitJobDriver.html
   *
   * @default - No job driver parameters specified.
   */
  readonly sparkSubmitJobDriver?: SparkSubmitJobDriver;
}

/**
 * A configuration specification to be used when provisioning virtual clusters,
 * which can include configurations for applications and software bundled with Amazon EMR on EKS.
 * A configuration consists of a classification, properties, and optional nested configurations.
 * A classification refers to an application-specific configuration file.
 * Properties are the settings you want to change in that file.
 */
export interface Configuration {

  /**
   * The classification within a configuration.
   * Length Constraints: Minimum length of 1. Maximum length of 1024.
   */
  readonly classification: string;

  /**
   * A list of additional configurations to apply within a configuration object.
   * Array Members: Maximum number of 100 items.
   *
   * @default - No other configurations
   */
  readonly configurations?: Configuration[];

  /**
   * A set of properties specified within a configuration classification.
   * Map Entries: Maximum number of 100 items.
   *
   * @default - No properties
   */
  readonly properties?: { [key: string]: string };
}

/**
 * A configuration for CloudWatch monitoring.
 * You can configure your jobs to send log information to CloudWatch Logs.
 */
export interface CloudWatchMonitoringConfiguration {

  /**
   * The name of the log group for log publishing.
   * Length Constraints: Minimum length of 1. Maximum length of 512.
   *
   * @default - No default log group name
   */
  readonly logGroupName: string;

  /**
   * The specified name prefix for log streams.
   * Length Constraints: Minimum length of 1. Maximum length of 256.
   *
   * @default - No default log stream prefix name
   */
  readonly logStreamNamePrefix?: string;
}

/**
 *  Amazon S3 configuration for monitoring log publishing.
 *  You can configure your jobs to send log information to Amazon S3.
 */
export interface S3MonitoringConfiguration {

  /**
   * Amazon S3 destination URI for log publishing.
   * Length Constraints: Minimum length of 1. Maximum length of 10280.
   *
   * @default - No default S3 Destination URI
   */
  readonly logUri: string;
}

/**
 * Configuration setting for monitoring.
 */
export interface MonitoringConfiguration {

  /**
   * A configuration for CloudWatch monitoring.
   * You can configure your jobs to send log information to CloudWatch Logs.
   *
   * @default - No CloudWatch monitoring configuration
   */
  readonly cloudWatchMonitoringConfig?: CloudWatchMonitoringConfiguration;

  /**
   * Monitoring configurations for the persistent application UI.
   * Valid values - ENABLED | DISABLED
   *
   * @default - No persistent application UI configuration
   */
  readonly persistentAppUI?: string;

  /**
   *  Amazon S3 configuration for monitoring log publishing.
   *  You can configure your jobs to send log information to Amazon S3.
   *
   * @default - No S3 monitoring configuration
   */
  readonly s3MonitoringConfig?: S3MonitoringConfiguration;
}

/**
 * A configuration specification to be used to override existing configurations.
 */
export interface ConfigurationOverrides {

  /**
   * The configurations for the application running by the job run.
   * Maximum of 100 items
   *
   * @see https://docs.aws.amazon.com/emr-on-eks/latest/APIReference/API_Configuration.html
   *
   * @default - No application config
   */
  readonly applicationConfig?: Configuration[];

  /**
   * The configurations for monitoring.
   *
   * @see https://docs.aws.amazon.com/emr-on-eks/latest/APIReference/API_MonitoringConfiguration.html
   *
   * @default - No monitoring configurations
   */
  readonly monitoringConfig?: MonitoringConfiguration;
}
