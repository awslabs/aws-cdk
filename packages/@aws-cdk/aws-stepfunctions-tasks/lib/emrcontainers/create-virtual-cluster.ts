// eslint-disable-next-line import/no-extraneous-dependencies
import * as iam from '@aws-cdk/aws-iam';
import * as sfn from '@aws-cdk/aws-stepfunctions';
import { Stack } from '@aws-cdk/core';
import { Construct } from 'constructs';
import { integrationResourceArn, validatePatternSupported } from '../private/task-utils';

/**
 * Enum for supported types of Container Providers
 */
export enum ContainerProviderTypes {

  /**
   * Supported container provider type is an EKS Cluster
   */
  EKS = 'EKS',
}

/**
 * The information about the EKS cluster.
 */
export interface EksInfo {

  /**
   * The namespaces of the EKS cluster.
   * Length Constraints: Minimum length of 1. Maximum length of 63.
   *
   * @default - No namespace
   */
  readonly namespace?: string;
}

/**
 * The information about the container used for a job run or a managed endpoint.
 */
export interface ContainerInfo {

  /**
   * The information about the EKS cluster.
   *
   * @default - No EKS info
   */
  readonly eksInfo?: EksInfo;
}

/**
 * The information about the container provider.
 */
export interface ContainerProvider {

  /**
   * The ID of the container cluster.
   *
   * @default - No Cluster Id
   */
  readonly id: string;

  /**
   * The information about the container cluster.
   *
   * @default - No container info
   */
  readonly info?: ContainerInfo;

  /**
   * The type of the container provider.
   * EKS is the only supported type as of now.
   */
  readonly type: string;
}

/**
 * The props for a EMR Containers CreateVirtualCluster Task.
 */
export interface EmrContainersCreateVirtualClusterProps extends sfn.TaskStateBaseProps {

  /**
   * Name of the specified virtual cluster.
   *
   */
  readonly name: string;

  /**
   * The container provider of the virtual cluster
   *
   * @see https://docs.aws.amazon.com/emr-on-eks/latest/APIReference/API_ContainerProvider.html
   *
   * @default - No container provider
   */
  readonly containerProvider: ContainerProvider;

  /**
   * The tags assigned to the virtual cluster
   *
   * @default {}
   */
  readonly tags?: { [key: string]: string };
}

/**
 * Creates a Virtual Cluster from a EKS cluster in a Task State
 *
 * @see https://docs.amazonaws.cn/en_us/step-functions/latest/dg/connect-emr-eks.html
 */
export class EmrContainersCreateVirtualCluster extends sfn.TaskStateBase {

  private static readonly SUPPORTED_INTEGRATION_PATTERNS: sfn.IntegrationPattern[] = [
    sfn.IntegrationPattern.REQUEST_RESPONSE,
  ];

  protected readonly taskMetrics?: sfn.TaskMetricsConfig;
  protected readonly taskPolicies?: iam.PolicyStatement[];

  private readonly integrationPattern: sfn.IntegrationPattern;

  constructor(scope: Construct, id: string, private readonly props: EMRContainersCreateVirtualClusterProps) {
    super(scope, id, props);
    this.integrationPattern = props.integrationPattern ?? sfn.IntegrationPattern.REQUEST_RESPONSE;
    validatePatternSupported(this.integrationPattern, EMRContainersCreateVirtualCluster.SUPPORTED_INTEGRATION_PATTERNS);

    this.taskPolicies = this.createPolicyStatements();
  }

  /**
   * @internal
   */
  protected _renderTask(): any {
    return {
      Resource: integrationResourceArn('emr-containers', 'createVirtualCluster', this.integrationPattern),
      Parameters: sfn.FieldUtils.renderObject({
        Name: this.props.name,
        ContainerProvider: {
          Id: this.props.containerProvider.id,
          Info: {
            EksInfo: {
              Namespace: this.props.containerProvider.info?.eksInfo?.namespace,
            },
          },
          Type: this.props.containerProvider.type,
        },
        Tags: this.props.tags,
      }),
    };
  };

  private createPolicyStatements(): iam.PolicyStatement[] {
    return [
      new iam.PolicyStatement({
        resources: ['*'],
        actions: ['emr-containers:CreateVirtualCluster'],
      }),
      new iam.PolicyStatement({
        resources: ['*'],
        actions: ['iam:CreateServiceLinkedRole'],
        conditions: {
          stringLike: {
            'iam:AWSServiceName': 'emr-containers.amazonaws.com',
          },
        },
      }),
    ];
  }
}
