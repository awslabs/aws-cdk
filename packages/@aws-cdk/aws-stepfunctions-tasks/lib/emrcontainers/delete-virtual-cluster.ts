import * as iam from '@aws-cdk/aws-iam';
import * as sfn from '@aws-cdk/aws-stepfunctions';
import { Stack } from '@aws-cdk/core';
import { Construct } from 'constructs';
import { integrationResourceArn, validatePatternSupported } from '../private/task-utils';

/**
 * The props for a EMR Containers DeleteVirtualCluster Task.
 */
export interface EMRContainersDeleteVirtualClusterProps extends sfn.TaskStateBaseProps {
  /**
   * The ID of the virtual cluster that will be deleted.
   */
  readonly virtualClusterId: sfn.TaskInput;
}

/**
 * Deletes a EMR Containers virtual cluster as a Task.
 *
 * @see https://docs.amazonaws.cn/en_us/step-functions/latest/dg/connect-emr-eks.html
 */
export class EMRContainersDeleteVirtualCluster extends sfn.TaskStateBase {

  private static readonly SUPPORTED_INTEGRATION_PATTERNS: sfn.IntegrationPattern[] = [
    sfn.IntegrationPattern.REQUEST_RESPONSE,
    sfn.IntegrationPattern.RUN_JOB,
  ];

  protected readonly taskMetrics?: sfn.TaskMetricsConfig;
  protected readonly taskPolicies?: iam.PolicyStatement[];

  private readonly integrationPattern: sfn.IntegrationPattern;

  constructor(scope: Construct, id: string, private readonly props: EMRContainersDeleteVirtualClusterProps) {
    super(scope, id, props);
    this.integrationPattern = props.integrationPattern ?? sfn.IntegrationPattern.RUN_JOB;

    validatePatternSupported(this.integrationPattern, EMRContainersDeleteVirtualCluster.SUPPORTED_INTEGRATION_PATTERNS);

    this.taskPolicies = this.createPolicyStatements();
  }

  /**
   * @internal
   */
  protected _renderTask(): any {
    return {
      Resource: integrationResourceArn('emr-containers', 'deleteVirtualCluster', this.integrationPattern),
      Parameters: sfn.FieldUtils.renderObject({
        Id: this.props.id,
      }),
    };
  };

  private createPolicyStatements(): iam.PolicyStatement[] {
    let iamActions: string[] | undefined;

    if (this.integrationPattern === sfn.IntegrationPattern.REQUEST_RESPONSE) {
      iamActions = ['emr-containers:DeleteVirtualCluster'];
    } else if (this.integrationPattern === sfn.IntegrationPattern.RUN_JOB) {
      iamActions = [
        'emr-containers:ListVirtualClusters',
        'emr-containers:DescribeVirtualCluster',
        'emr-containers:DeleteVirtualCluster',
      ];
    }

    return [new iam.PolicyStatement({
      resources: ['*'],
      actions: iamActions,
    })];
  }
}
