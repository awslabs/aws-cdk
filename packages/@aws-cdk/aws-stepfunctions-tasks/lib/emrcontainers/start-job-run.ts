import * as iam from '@aws-cdk/aws-iam';
import * as sfn from '@aws-cdk/aws-stepfunctions';
import { Construct } from 'constructs';
import { integrationResourceArn, validatePatternSupported } from '../private/task-utils';
import { JobDriver, ConfigurationOverrides } from './base-types';

/**
 * The props for a EMR Containers StartJobRun Task.
 */
export interface EMRContainersStartJobRunProps extends sfn.TaskStateBaseProps {

  /**
   * The virtual cluster ID for which the job run request is submitted.
   */
  readonly clusterId: string;

  /**
   * The name of the job run.
   *
   * @default - No job name
   */
  readonly jobName?: string;

  /**
   * The execution role ARN for the job run.
   *
   * @default - No execution arn
   */
  readonly executionRoleArn: string;

  /**
   * The Amazon EMR release version to use for the job run.
   */
  readonly releaseLabel: string;

  /**
   * The job driver for the job run.
   *
   * @see https://docs.aws.amazon.com/emr-on-eks/latest/APIReference/API_JobDriver.html
   *
   * @default - No job driver
   */
  readonly jobDriver: JobDriver;

  /**
   * A configuration specification to be used to override existing configurations.
   *
   * @see https://docs.aws.amazon.com/emr-on-eks/latest/APIReference/API_ConfigurationOverrides.html
   *
   * @default - No configuration override
   */
  readonly configurationOverrides?: ConfigurationOverrides;

  /**
   * The tags assigned to job runs.
   *
   * @default - None
   */
  readonly tags?: { [key: string]: string };
}

/**
 * Starts a job run. A job run is a unit of work, such as a Spark jar, PySpark script, or SparkSQL query,
 * that you submit to Amazon EMR on EKS.
 *
 * @see https://docs.amazonaws.cn/en_us/step-functions/latest/dg/connect-emr-eks.html
 */
export class EMRContainersStartJobRun extends sfn.TaskStateBase {
  private static readonly SUPPORTED_INTEGRATION_PATTERNS: sfn.IntegrationPattern[] = [
    sfn.IntegrationPattern.REQUEST_RESPONSE,
    sfn.IntegrationPattern.RUN_JOB,
  ];

  protected readonly taskMetrics?: sfn.TaskMetricsConfig;
  protected readonly taskPolicies?: iam.PolicyStatement[];

  private readonly integrationPattern: sfn.IntegrationPattern;

  constructor(scope: Construct, id: string, private readonly props: EMRContainersStartJobRunProps) {
    super(scope, id, props);
    this.integrationPattern = props.integrationPattern ?? sfn.IntegrationPattern.RUN_JOB;

    validatePatternSupported(this.integrationPattern, EMRContainersStartJobRun.SUPPORTED_INTEGRATION_PATTERNS);
  }

  /**
   * @internal
   *
   * TODO: ALL of this needs work
   */
  protected _renderTask(): any {
    return {
      Resource: integrationResourceArn('emr-containers', 'startJobRun', this.integrationPattern),
      Parameters: sfn.FieldUtils.renderObject({
        VirtualClusterId: this.props.clusterId,
        JobName: this.props.jobName,
        ExecutionRoleArn: this.props.executionRoleArn,
        ReleaseLabel: this.props.releaseLabel,
        JobDriver: this.props.jobDriver,
        ConfigurationOverrides: this.props.configurationOverrides,
        Tags: this.props.tags,
      }),
    };
  };
}
