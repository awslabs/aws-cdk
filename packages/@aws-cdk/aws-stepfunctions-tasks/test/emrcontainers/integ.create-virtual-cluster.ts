import * as sfn from '@aws-cdk/aws-stepfunctions';
import * as cdk from '@aws-cdk/core';
import { ContainerProviderTypes, EMRContainersCreateVirtualCluster } from '../../lib/emrcontainers/create-virtual-cluster';

/**
 * Stack verification steps:
 * Everything in the link below besides the last step must be setup before running the state machine.
 * @see https://docs.aws.amazon.com/emr/latest/EMR-on-EKS-DevelopmentGuide/setting-up.html
 *
 * aws stepfunctions start-execution --state-machine-arn <deployed state machine arn> : should return execution arn
 * aws stepfunctions describe-execution --execution-arn <exection-arn generated before> : should return status as SUCCEEDED
 */

const app = new cdk.App();
const stack = new cdk.Stack(app, 'aws-stepfunctions-tasks-emr-containers-create-virtual-cluster-integ');

const createVirtualClusterJob = new EMRContainersCreateVirtualCluster(stack, 'Create a Virtual Cluster Job', {
  name: 'emr-containers-test-cluster',
  containerProvider: {
    id: 'test-eks',
    info: {
      eksInfo: {
        namespace: 'kube-system',
      },
    },
    type: ContainerProviderTypes.EKS,
  },
  integrationPattern: sfn.IntegrationPattern.REQUEST_RESPONSE,
});

const chain = sfn.Chain.start(createVirtualClusterJob);

const sm = new sfn.StateMachine(stack, 'StateMachine', {
  definition: chain,
  timeout: cdk.Duration.seconds(30),
});

new cdk.CfnOutput(stack, 'stateMachineArn', {
  value: sm.stateMachineArn,
});


app.synth();
