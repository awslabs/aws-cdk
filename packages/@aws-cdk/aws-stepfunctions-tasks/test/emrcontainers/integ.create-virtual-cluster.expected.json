{
  "Resources": {
      "StateMachineRoleB840431D": {
          "Type": "AWS::IAM::Role",
          "Properties": {
              "AssumeRolePolicyDocument": {
                  "Statement": [
                      {
                          "Action": "sts:AssumeRole",
                          "Effect": "Allow",
                          "Principal": {
                              "Service": {
                                  "Fn::Join": [
                                      "",
                                      [
                                          "states.",
                                          {
                                              "Ref": "AWS::Region"
                                          },
                                          ".amazonaws.com"
                                      ]
                                  ]
                              }
                          }
                      }
                  ],
                  "Version": "2012-10-17"
              }
          },
          "Metadata": {
              "aws:cdk:path": "aws-stepfunctions-tasks-emr-containers-create-virtual-cluster-integ/StateMachine/Role/Resource"
          }
      },
      "StateMachineRoleDefaultPolicyDF1E6607": {
          "Type": "AWS::IAM::Policy",
          "Properties": {
              "PolicyDocument": {
                  "Statement": [
                      {
                          "Action": "emr-containers:CreateVirtualCluster",
                          "Effect": "Allow",
                          "Resource": {
                              "Fn::Join": [
                                  "",
                                  [
                                      "arn:",
                                      {
                                          "Ref": "AWS::Partition"
                                      },
                                      ":emr-containers:",
                                      {
                                          "Ref": "AWS::Region"
                                      },
                                      ":",
                                      {
                                          "Ref": "AWS::AccountId"
                                      },
                                      ":virtual-cluster/emr-containers-test-cluster"
                                  ]
                              ]
                          }
                      }
                  ],
                  "Version": "2012-10-17"
              },
              "PolicyName": "StateMachineRoleDefaultPolicyDF1E6607",
              "Roles": [
                  {
                      "Ref": "StateMachineRoleB840431D"
                  }
              ]
          },
          "Metadata": {
              "aws:cdk:path": "aws-stepfunctions-tasks-emr-containers-create-virtual-cluster-integ/StateMachine/Role/DefaultPolicy/Resource"
          }
      },
      "StateMachine2E01A3A5": {
          "Type": "AWS::StepFunctions::StateMachine",
          "Properties": {
              "RoleArn": {
                  "Fn::GetAtt": [
                      "StateMachineRoleB840431D",
                      "Arn"
                  ]
              },
              "DefinitionString": {
                  "Fn::Join": [
                      "",
                      [
                          "{\"StartAt\":\"Create a Virtual Cluster Job\",\"States\":{\"Create a Virtual Cluster Job\":{\"End\":true,\"Type\":\"Task\",\"Resource\":\"arn:",
                          {
                              "Ref": "AWS::Partition"
                          },
                          ":states:::emr-containers:createVirtualCluster\",\"Parameters\":{\"Name\":\"emr-containers-test-cluster\",\"ContainerProvider\":{\"Id\":\"test-eks\",\"Info\":{\"EksInfo\":{\"Namespace\":\"kube-system\"}},\"Type\":\"EKS\"}}}},\"TimeoutSeconds\":30}"
                      ]
                  ]
              }
          },
          "DependsOn": [
              "StateMachineRoleDefaultPolicyDF1E6607",
              "StateMachineRoleB840431D"
          ],
          "Metadata": {
              "aws:cdk:path": "aws-stepfunctions-tasks-emr-containers-create-virtual-cluster-integ/StateMachine/Resource"
          }
      },
      "CDKMetadata": {
          "Type": "AWS::CDK::Metadata",
          "Properties": {
              "Analytics": "v2:deflate64:H4sIAAAAAAAAE1VPzQrCMAx+Fu9dNz14ForHoUzwXroMs9V2NKkio+9uu1309P3lS0gjG9nsTvpNlemnejE+gFxurM0k1OAukefIQnlHHKLh4nVAPgYDheegR0bvkigrFmKYh+hMsahiTRPJc9vlMdboIJAKoBnuGDhqq2zMhSD+Wut1hlabR26UI786CdRPuXTertGKV2/RfIrcWEpJON+DHKl+7Y/ykD8cCbEK0TE+QXYbfgEtN7B1/AAAAA=="
          },
          "Metadata": {
              "aws:cdk:path": "aws-stepfunctions-tasks-emr-containers-create-virtual-cluster-integ/CDKMetadata/Default"
          },
          "Condition": "CDKMetadataAvailable"
      }
  },
  "Outputs": {
      "stateMachineArn": {
          "Value": {
              "Ref": "StateMachine2E01A3A5"
          }
      }
  },
  "Conditions": {
      "CDKMetadataAvailable": {
          "Fn::Or": [
              {
                  "Fn::Or": [
                      {
                          "Fn::Equals": [
                              {
                                  "Ref": "AWS::Region"
                              },
                              "af-south-1"
                          ]
                      },
                      {
                          "Fn::Equals": [
                              {
                                  "Ref": "AWS::Region"
                              },
                              "ap-east-1"
                          ]
                      },
                      {
                          "Fn::Equals": [
                              {
                                  "Ref": "AWS::Region"
                              },
                              "ap-northeast-1"
                          ]
                      },
                      {
                          "Fn::Equals": [
                              {
                                  "Ref": "AWS::Region"
                              },
                              "ap-northeast-2"
                          ]
                      },
                      {
                          "Fn::Equals": [
                              {
                                  "Ref": "AWS::Region"
                              },
                              "ap-south-1"
                          ]
                      },
                      {
                          "Fn::Equals": [
                              {
                                  "Ref": "AWS::Region"
                              },
                              "ap-southeast-1"
                          ]
                      },
                      {
                          "Fn::Equals": [
                              {
                                  "Ref": "AWS::Region"
                              },
                              "ap-southeast-2"
                          ]
                      },
                      {
                          "Fn::Equals": [
                              {
                                  "Ref": "AWS::Region"
                              },
                              "ca-central-1"
                          ]
                      },
                      {
                          "Fn::Equals": [
                              {
                                  "Ref": "AWS::Region"
                              },
                              "cn-north-1"
                          ]
                      },
                      {
                          "Fn::Equals": [
                              {
                                  "Ref": "AWS::Region"
                              },
                              "cn-northwest-1"
                          ]
                      }
                  ]
              },
              {
                  "Fn::Or": [
                      {
                          "Fn::Equals": [
                              {
                                  "Ref": "AWS::Region"
                              },
                              "eu-central-1"
                          ]
                      },
                      {
                          "Fn::Equals": [
                              {
                                  "Ref": "AWS::Region"
                              },
                              "eu-north-1"
                          ]
                      },
                      {
                          "Fn::Equals": [
                              {
                                  "Ref": "AWS::Region"
                              },
                              "eu-south-1"
                          ]
                      },
                      {
                          "Fn::Equals": [
                              {
                                  "Ref": "AWS::Region"
                              },
                              "eu-west-1"
                          ]
                      },
                      {
                          "Fn::Equals": [
                              {
                                  "Ref": "AWS::Region"
                              },
                              "eu-west-2"
                          ]
                      },
                      {
                          "Fn::Equals": [
                              {
                                  "Ref": "AWS::Region"
                              },
                              "eu-west-3"
                          ]
                      },
                      {
                          "Fn::Equals": [
                              {
                                  "Ref": "AWS::Region"
                              },
                              "me-south-1"
                          ]
                      },
                      {
                          "Fn::Equals": [
                              {
                                  "Ref": "AWS::Region"
                              },
                              "sa-east-1"
                          ]
                      },
                      {
                          "Fn::Equals": [
                              {
                                  "Ref": "AWS::Region"
                              },
                              "us-east-1"
                          ]
                      },
                      {
                          "Fn::Equals": [
                              {
                                  "Ref": "AWS::Region"
                              },
                              "us-east-2"
                          ]
                      }
                  ]
              },
              {
                  "Fn::Or": [
                      {
                          "Fn::Equals": [
                              {
                                  "Ref": "AWS::Region"
                              },
                              "us-west-1"
                          ]
                      },
                      {
                          "Fn::Equals": [
                              {
                                  "Ref": "AWS::Region"
                              },
                              "us-west-2"
                          ]
                      }
                  ]
              }
          ]
      }
  }
}